### ISSUE TITLE
Bootloader transfer

### ISSUE THREAD

**Issue by MattTraina** [Altronic-LLC/Altronic-Public-Files#18]
Integration of bootloader to DCPM to run application that is already developed.

**Comment by Aaron-Phifer** [comment]
Bootloader for CIU is located here: https://github.com/Altronic-LLC/NGI-5000-CIU-bootloader
Application for CIU is located here: https://github.com/Altronic-LLC/NGI-5000-CIU/tree/NGI-5000_Rev3_Board
Go code that can run on the DCPM for sending bin file: https://github.com/Altronic-LLC/dcpm_ciu_bootloader

**Comment by MattTraina** [comment]
@lm1099 there is currently a button for the ST to perform the bootloader, and the CM4 on the logic requires you to run something from the desktop. It would be really good if there was a way to trigger it from the logic screen in the AWI with a button. I believe there is something triggered on the ACM from a register that is written to once the IRIS tool sees it.

**Comment by lm1099** [comment]
@worthb0123 As I understand it, the AWI will kick off the CIU update process by writing to a specific register that the iris_go program will note, which will then send the new CIU program to the DCPM over SCP and forward the register write to the scanner. When the scanner sees a specific register value, it will start the programming process for all connected CIUs using the file that was copied over. It will also update a modbus register holding a progress indicator.

Are the registers that we should use already in the spreadsheets for the DCPM and the Logic module?

**Comment by MattTraina** [comment]
In the 2k_min spreadsheet https://docs.google.com/spreadsheets/d/1f1OWV0nSlkCpmUf96dhUbYpOnmwyNIPt8zAGPdEB5vc/edit?gid=0#gid=0

register 999 starts the bootloader. Entering the number 27183 (in the case of the AWI we push a button that sends that value) the iris go will initiate the bootloader process. @Aaron-Phifer do you know what happens if there is a number entered that it is not looking for? We could make it so that 999 is the bootloader register and the value entered and seen would make different boot loaders  kick off. Otherwise we must create and compile for different bootloader registers. This could be okay as well.

**Comment by MattTraina** [comment]
![Image](https://github.com/Altronic-LLC/Altronic-Public-Files/assets/56164635/4b8fbbc5-a71e-43b1-b854-0b432c8981f5)

**Comment by MattTraina** [comment]
It looks like we could reuse this framework, but instead of opening the file we could send the file to the DCPM. And then logic in the DCPM can trigger it's process once receiving the file. It could be such that the DCPM and Logic CM4 get the file transferred and once that happens another register could be set that the DCPM also sees so it starts only when the transfer was successful. As of now, I am thinking of that same 999 register with a different value entered. I've added 10001 as the file has transfered coil. This could be removed if not needed, but this is what I am planning for at the moment. 

![Image](https://github.com/Altronic-LLC/Altronic-Public-Files/assets/56164635/d3fa7691-ee38-4054-8962-3882e7e6fff7)

**Comment by worthb0123** [comment]
Matt, I was right in the middle of adding the appropriate registers. I'll be responding with the new register numbers shortly. We need to keep any DCPM/CIU register numbers above 9000.

**Comment by worthb0123** [comment]
@lm1099 @MattTraina 
I have created a few new registers for controlling and monitoring the CIU bootloader progress.

In the 40000 block there is a new register 9997. The label is "CIU Bootloader Start". On the ST bootloader we write the value 27183 to start the bootloader process. I think this is just a random value that Aaron and I picked. The reasoning was to prevent an errant write to this register to start the bootloader process.

As Matt mentioned we could write the values between 1 and 40 to select a single CIU to be written. I know that for our initial implementation we said we would just write to all connected CIU's. I suggest that for now we just write "999" to this register in order to start the bootloader process for all CIU's. In the future we might use an additional register to select the individual CIU to be written to, but we still write "999" to register 9997 to kick off the process.

The other new registers are for reading status, so they are 30000 registers:
9997: CIU Bootloader State - As with the ST bootloader, this register can have text enumerations to display the state of the bootloader

9996: CIU Bootloader Progress - As with the ST bootloader, this register will have a value between 0 - 100 to represent the percentage of progress. This may need to be re-designed because we will have multiple CIU's being updated at the same time. We could have 40 multiple registers and each could have a progress percentage. For now, we could use the single register to indicate the percentage of the number of CIU's that have been successfully programmed.

9995: CIU Bootloader Version - As with the ST bootloader, this register will have a 16 bit integer that stores a firmware version number. We will need to think through how to extract this value from the firmware file and push it into this register.

Again, we may need to add additional register or revisit some of these ideas, but this should be enough to keep moving forward.

**Comment by lm1099** [comment]
@worthb0123 Have the new register updates been written to the spreadsheets yet? I refreshed the data using iris_tool, but I wasn't able to find the new registers in the coredata_dcpm JSON file.

```
  "000066": {
    "block": "40000",
    "register": 9533,
    "key": "KEYCMD_G",
    "factor": 1,
    "offset": 0,
    "type": "UINT16",
    "defaultValue": 0,
    "regLen": 1,
    "index": 66
  },
  "000067": {
    "block": "40000",
    "register": 10001,
    "key": "REG40001_C1",
    "factor": 1,
    "offset": 0,
    "type": "UINT16",
    "defaultValue": 0,
    "regLen": 1,
    "index": 67
  },
```

```
  "000000": {
    "block": "30000",
    "register": 1,
    "key": "TEST30001",
    "factor": 1,
    "offset": 0,
    "type": "UINT16",
    "defaultValue": 0,
    "regLen": 1,
    "index": 0
  },
  "000001": {
    "block": "30000",
    "register": 10001,
    "key": "SBITS1_C1",
    "factor": 1,
    "offset": 0,
    "type": "UINT16",
    "defaultValue": 0,
    "regLen": 1,
    "index": 1
  },
```

**Comment by worthb0123** [comment]
@lm1099 did you run the refresh script before you ran the iris tool?

**Comment by lm1099** [comment]
@worthb0123 Yes, I ran refresh.sh right before running the iris_tool.go tool.

**Comment by worthb0123** [comment]
@lm1099 and you are selecting the ngi5000 2kmin device?

**Comment by worthb0123** [comment]
@lm1099 to answer you original question, yes the new registers had already been added to the spreadsheet.

**Comment by lm1099** [comment]
@worthb0123 I was selecting the DCPM device since I thought registers >9000 were only contained in that file, but I downloaded the coredata file for NGI-5000-LOGIC-2K_min as well and wasn't able to find them there either. Just to clarify, I should be looking for (40000, 9997), (30000, 9997), (30000, 9996), (30000, 9995), right?

**Comment by worthb0123** [comment]
@lm1099 First, run refresh.sh. This will download the latest device-profiles.json. You can edit this file and search for 9997. You should find a match in the 30000 registers and also in the 40000 registers.

Next, do "go run tool.go" and be sure to select the NGI-5000-LOGIC-2K_min as the device. When the tool is finished running, there will be new files in the "output" folder. The file you want is coredata_dcpm.json. Just copy this file to the dcpm_scanner folder.

You should be able to search for "9997" inside of dcpm_scanner.json and see the new registers in the 30000 and 40000 blocks.

**Comment by lm1099** [comment]
@worthb0123 Got it, thanks! I didn't realize that NGI-5000-LOGIC-2K_min generated both coredata.json and coredata_dcpm.json. I can see all of the registers now
